pkg/kademlia/pslice/pslice_test.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pslice_test
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/kademlia/pslice"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology/test"
+)
+
+// TestShallowestEmpty tests that ShallowestEmpty functionality works correctly.
+func TestShallowestEmpty(t *testing.T) {
+	var (
+		ps    = pslice.New(16)
+		base  = test.RandomAddress()
+		peers = make([]swarm.Address, 16)
+	)
+
+	for i := 0; i < 16; i++ {
+		a := test.RandomAddressAt(base, i)
+		peers[i] = a
+	}
+
+	for i, v := range peers {
+		ps.Add(v, uint8(i))
+		sd, none := ps.ShallowestEmpty()
+		if i == 15 {
+			if !none {
+				t.Fatal("expected last bin to be empty and return no empty bins")
+			}
+		} else {
+			if sd != uint8(i+1) {
+				t.Fatalf("expected shallow empty bin to be %d but got %d", i+1, sd)
+			}
+			if none {
+				t.Fatal("got no empty bins but wanted some")
+			}
+		}
+	}
+
+	for _, tc := range []struct {
+		rmPO  int
+		expSE uint8
+	}{
+		{rmPO: 3,
+			expSE: 3,
+		}, {
+			rmPO:  1,
+			expSE: 1,
+		}, {
+			rmPO:  10,
+			expSE: 1,
+		}, {
+			rmPO:  15,
+			expSE: 1,
+		}, {
+			rmPO:  14,
+			expSE: 1,
+		}, {
+			rmPO:  0,
+			expSE: 0,
+		},
+	} {
+		po := uint8(swarm.Proximity(base.Bytes(), peers[tc.rmPO].Bytes()))
+		ps.Remove(peers[tc.rmPO], po)
+
+		sd, none := ps.ShallowestEmpty()
+		if sd != tc.expSE || none {
```

Comment:
should this test perhaps be performed with _more_ than one peer in each bin, too?

Reply:
yes but then i have to rewrite this entire test from the beginning. which would suck badly. i'll create a different test for this then

Reply:
`cat | /dev/debt`